### ISSUE TITLE
How to train a model, use mulit-host data parallel?

### ISSUE THREAD

**Issue by matrix97317** [young-geng/scalax#8]
@young-geng Hi, Can you provide a demo about training model using mulit-host data parallel method?

**Comment by young-geng** [comment]
If you want to see a full fledged codebase with support for multi-host LLM training with various kinds of parallelism implemented with scalax, you can checkout my [mintext](https://github.com/young-geng/mintext) repo.

**Comment by matrix97317** [comment]
Thank you for your reply. However, I only see parameter splitting mentioned, but I don't see any operations related to parameter or gradient synchronization between different hosts.  Here is a demo I wrote using Scalax.
```
 jax.distributed.initialize(f'{master_addr}:10000',world_size, rank)
 mesh = MeshShardingHelper([-1], ['fsdp'])
 model_sharding_rule = FSDPShardingRule(fsdp_axis_name='fsdp')
 ptrain_step = mesh.sjit(functools.partial(train_step, config),
        in_shardings=(PartitionSpec(),model_sharding_rule, PartitionSpec()),
        out_shardings=(model_sharding_rule, PartitionSpec()),
        args_sharding_constraint=(PartitionSpec(),model_sharding_rule, PartitionSpec()),
        donate_argnums=(0,),  # No need to preserve the old model
    )
 train_state, info = ptrain_step(train_rng, train_state, batch)
       
 traverse_all_reduce_params(train_state.params)

# Add parameter synchronization across hosts.  Without it, the loss will converge more slowly. However, for large models, it is likely to cause GPU out-of-memory (OOM) errors.

```

The architecture of my distributed system is structured as follows:

Host-0            ｜    Host-1
D0 D0 D0 D0 ｜ D1 D1 D1 D1
M0 M1 M2 M3 ｜M0 M1 M2 M3  （FSDP split parameters to different GPU）
G0 G1  G2 G3 ｜G0 G1 G2 G3       （indicate different GPU）
all-reduce ｜ all-reduce

I expect the architecture to be like this:

Host-0            ｜    Host-1
D1 D2 D3 D4 ｜ D5 D6 D7 D8
M0 M1 M2 M3 ｜M4 M5 M6 M7  （FSDP split parameters to different GPU，can auto all-reduce）
G0 G1  G2 G3 ｜G0 G1 G2 G3       （indicate different GPU）

But jax can not sync paramters across GPUs

**Comment by young-geng** [comment]
The code you show should perform reduction overall all hosts. Do you see all devices on all hosts if you call `jax.devices()` after jax distributed initialization?

BTW, you should also shard your data using the FSDP axis. You can do it either by putting `PartitionSpec('fsdp')` in `in_shardings` or `args_sharding_constraint` of sjit.

**Comment by matrix97317** [comment]
@young-geng Thank you very much.I have use this structure:

Host-0 ｜ Host-1
Dg0-1 Dg0-2 Dg0-3 Dg0-4 ｜Dg1-1 Dg1-2 Dg1-3 Dg1-4 (args_sharding_constraint(PartitionSpec('fsdp'))
M0 M1 M2 M3 ｜M0 M1 M2 M3 （FSDP split parameters to different GPU）
G0 G1 G2 G3 ｜G0 G1 G2 G3 （indicate different GPU）
all-reduce ｜ all-reduce
